shell/platform/android/io/flutter/plugin/editing/SpellCheckPlugin.java

```diff
@@ -0,0 +1,113 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugin.editing;
+
+import android.content.Context;
+import android.view.textservice.SentenceSuggestionsInfo;
+import android.view.textservice.SpellCheckerSession;
+import android.view.textservice.SuggestionsInfo;
+import android.view.textservice.TextInfo;
+import android.view.textservice.TextServicesManager;
+import androidx.annotation.NonNull;
+import androidx.annotation.VisibleForTesting;
+import io.flutter.embedding.engine.systemchannels.SpellCheckChannel;
+import java.util.ArrayList;
+import java.util.Locale;
+
+/** Android implementation of the spell check plugin. */
+public class SpellCheckPlugin implements SpellCheckerSession.SpellCheckerSessionListener {
+
+  @NonNull private final Context mContext;
+  @NonNull private final SpellCheckChannel mSpellCheckChannel;
+  @NonNull private final TextServicesManager tsm;
+  private SpellCheckerSession mSpellCheckerSession;
+
+  @VisibleForTesting @NonNull
+  final SpellCheckChannel.SpellCheckMethodHandler mSpellCheckMethodHandler;
+
+  public SpellCheckPlugin(@NonNull Context context, @NonNull SpellCheckChannel spellCheckChannel) {
+    mContext = context;
+    mSpellCheckChannel = spellCheckChannel;
+    tsm = (TextServicesManager) mContext.getSystemService(Context.TEXT_SERVICES_MANAGER_SERVICE);
+
+    mSpellCheckMethodHandler =
+        new SpellCheckChannel.SpellCheckMethodHandler() {
+          @Override
+          public void initiateSpellCheck(String locale, String text) {
+            performSpellCheck(locale, text);
+          }
+        };
+
+    mSpellCheckChannel.setSpellCheckMethodHandler(mSpellCheckMethodHandler);
+  }
+
+  public void destroy() {
+    mSpellCheckChannel.setSpellCheckMethodHandler(null);
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
+    }
+  }
+
+  /** Calls on the Android spell check API to spell check specified text. */
+  public void performSpellCheck(String locale, String text) {
+    String[] localeCodes = locale.split("-");
+    Locale parsedLocale;
+
+    if (localeCodes.length == 3) {
+      parsedLocale = new Locale(localeCodes[0], localeCodes[1], localeCodes[2]);
+    } else if (localeCodes.length == 2) {
+      parsedLocale = new Locale(localeCodes[0], localeCodes[1]);
+    } else {
+      parsedLocale = new Locale(localeCodes[0]);
+    }
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
```

Comment:
Is this related to the bug we were looking at where the spell checker would give inconsistent results?  So it seems like we start a new session every time the text changes.  I think that's right by briefly looking at the android spell check docs.  It seems like they want you to keep a session alive if you want to spell check multiple subsets of a larger piece of text?  (Probably not a concern of Flutter's right now)

Reply:
Yeah, this is related to that bug. I do agree that this agrees with the flow in [this diagram](https://developer.android.com/guide/topics/text/spell-checker-framework#SpellCheckClient) if you are interacting with the Android spell check service directly.

When it comes to larger pieces of text, we can definitely play around with this! Although, by the time this becomes a concern for Flutter, it would probably be more preferable to no longer be spell checking the entirety of the inputted text each time anyways.